Rc1

I-mikan-I opened the pull request:
@jenniferhellar
As mentioned, I updated the DSP suite to use the same toolchain for building/executing as Embench-IoT.
There's a README for the Arm board with invocation options in the example directory.
Can you try if recording new results works for you with this toolchain? I already pushed size results to the baseline-dir, but don't have the board right now to get the speed baseline.

All you should need is python, scons, and an arm compiler collection to use the toolchain.

Let me know if there's any issues!

I-mikan-I committed:
Moving new IoT toolcahain to DSP: Directory Setup

I-mikan-I committed:
Delete legacy files

I-mikan-I committed:
Fix typo

I-mikan-I committed:
Undo syscall interface changes

jenniferhellar committed:
adding lms_f32 test

jenniferhellar committed:
minor fix to align naming convention

I-mikan-I commented:
@jenniferhellar
I merged your upstream changes, copied parts of the IoT documentation over in `/doc`, and kept the generator scripts (for data) in `benchmarks/....py`.

The `examples/` directory contains any board specific data and compiler/linker flags for building and running the suite for a given target. For instance, the `stm32f4discovery` directory has a makefile for the BSP and the required commands in README.md.

I don't have the stm32f4 on hand currently. If you do, could you try to run the benchmarks with the `--absolute` and `--baseline-output` flags to generate our baseline .json, and copy it into `baseline-data/speed.json`?

Let me know if there's anything you think we're missing for the initial release. Right now it closely mirrors the IoT structure with minor exceptions, for example for the data-generating python scripts.

I also removed any licensing information from the Manual until we can check it with Jeremy.

jenniferhellar commented:
Hey, great work on this!  A couple things...

Can we rename the "benchmarks" directory to "golden-ref" or something like that?

And can we move the DSP source code out of the "support" directory into something like "dsplib"?  It's not support code, that's the actual implementation of the DSP library functions so it's confusing to have it under support/

Under dsplib, put the header files into a sub-directory include/ or inc/ and source files under source/ or src/.  I'd like people who are familiar with CMSIS to be able to navigate the same hierarchy: https://github.com/ARM-software/CMSIS-DSP/tree/main

<img width="276" alt="image" src="https://github.com/user-attachments/assets/7f160186-bc9a-4211-837b-e4f32e544b27" />

raysimar commented:
Sounds good to me. All the best,RaySent from my iPhone, forgive my tersenessOn Apr 19, 2025, at 1:27 PM, jenniferhellar ***@***.***> wrote:﻿
  Hey, great work on this!  A couple things...
Can we rename the "benchmarks" directory to "golden-ref" or something like that?
And can we move the DSP source code out of the "support" directory into something like "dsplib"?  It's not support code, that's the actual implementation of the DSP library functions so it's confusing to have it under support/
image.png (view on web)—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>



jenniferhellar left a comment (embench/embench-dsp#1)
Hey, great work on this!  A couple things...
Can we rename the "benchmarks" directory to "golden-ref" or something like that?
And can we move the DSP source code out of the "support" directory into something like "dsplib"?  It's not support code, that's the actual implementation of the DSP library functions so it's confusing to have it under support/
image.png (view on web)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

I-mikan-I committed:
Rename /benchmarks directory, move library code to /lib

I-mikan-I commented:
Thanks! I went ahead and changed the directory structure as you suggested.

jenniferhellar commented:
Thanks!

1. Looks like the baseline data for size doesn't have the lms results.  I got an error for that when i ran benchmark_size
2. Under lib/ let's split the headers from the source files to match CMSIS.  So have dsp/ and all the arm_<>.h files under lib/inc and all the other directories under lib/src
3. snr.c and snr.h should go under support/ since those are for measuring error.
4. I was able to run benchmark_speed 
`./benchmark_speed.py --builddir bd-arm-gcc-14.0.1-speed \
  --target-module=run_stm32f4-discovery \
  --gdb-command=arm-none-eabi-gdb --gsf=1 --cpu-mhz=16 --absolute --baseline-output`
  but the output doesn't make sense to me?  I attached it
  
[speed-2025-04-20-151839.log](https://github.com/user-attachments/files/19826902/speed-2025-04-20-151839.log)

I'm guessing it's doing some MHz normalization or something?  Maybe Paolo can take a look if he has a board.

I-mikan-I commented:
I think the output looks strange because we need to scale the benchmarks using more iterations. The numbers should represent milliseconds.

jenniferhellar commented:
Hmm I didn't implement or generate input/output data for multiple iterations.  Would it be easier to just change the scale to nanoseconds instead?

I-mikan-I commented:
Would there be an issue with just using a global for loop around the benchmarks?
Thats basically what we've been doing for IoT, with some volatile variables to ensure that the compiler cannot optimize the iterations away.

jenniferhellar commented:
Yes.  Some of the DSP functions (e.g. CFFT) operate in-place, corrupting the original input data.  Others (e.g. LMS) are adaptive, so the behavior changes over iterations even if you provide the same input data.  We could add more test framework around the benchmarks to avoid this (i.e. memcpy's to have a separate input buffer, reset/reinit the filters between iterations, etc.), but it's not as simple as a for loop.

I-mikan-I committed:
Update Size Baseline Measurements

I-mikan-I committed:
Change msecs to usecs

I-mikan-I committed:
Refactor lib/; Fix native example; Fix run_gdbserver_sim

I-mikan-I commented:
Hi, sorry for getting to this late.
I refactored lib/ according to the CMSIS layout. I also changed msecs to usecs in all scripts, which I hope is accurate enough, considering you at least got single digit millisecond counts on the log you uploaded.

I also found 1-2 minor issues in existing python scripts and fixed them.
Considering the paper is out by now, if everything works on your end I think we should merge the changes.
